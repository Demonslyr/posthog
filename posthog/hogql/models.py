from dataclasses import dataclass, field
from enum import IntEnum, StrEnum
from typing import Any, Optional, Union
from uuid import UUID
from zoneinfo import ZoneInfo

from posthog.models.property import PropertyGroup


class PropertyDefinitionType(IntEnum):
    EVENT = 1
    PERSON = 2
    GROUP = 3
    SESSION = 4


class PropertyType(StrEnum):
    String = "String"
    Boolean = "Boolean"
    Numeric = "Numeric"
    DateTime = "DateTime"
    Duration = "Duration"


@dataclass
class TeamDataClass:
    id: int
    project_id: int
    timezone: str = "UTC"
    test_account_filters: list[dict] = field(default_factory=list)
    path_cleaning_filters: Optional[list[dict]] = None

    @property
    def timezone_info(self) -> ZoneInfo:
        return ZoneInfo(self.timezone)

    def path_cleaning_filter_models(self) -> list[dict]:
        """Return path cleaning filter configuration"""
        return self.path_cleaning_filters or []


@dataclass
class UserDataClass:
    id: int
    email: str
    first_name: str = ""
    is_active: bool = True


@dataclass
class CohortDataClass:
    id: int
    name: Optional[str]
    team_id: int
    project_id: int
    deleted: bool = False
    filters: Optional[dict] = None
    is_static: bool = False

    @property
    def pk(self) -> int:
        return self.id

    @property
    def properties(self) -> PropertyGroup:
        """Convert cohort filters to PropertyGroup"""
        if not self.filters:
            return PropertyGroup(type="AND", values=[])
        
        # This would need the actual conversion logic from the Cohort model
        # For now, returning a basic structure
        return PropertyGroup(type="AND", values=self.filters.get("properties", []))


@dataclass
class ActionStepJSON:
    tag_name: Optional[str] = None
    text: Optional[str] = None
    text_matching: Optional[str] = None
    href: Optional[str] = None
    href_matching: Optional[str] = None
    selector: Optional[str] = None
    url: Optional[str] = None
    url_matching: Optional[str] = "contains"
    event: Optional[str] = None
    properties: Optional[list[dict]] = None


@dataclass
class ActionDataClass:
    id: int
    name: Optional[str]
    team_id: int
    project_id: int
    steps_json: list[dict] = field(default_factory=list)

    @property
    def pk(self) -> int:
        return self.id

    @property
    def steps(self) -> list[ActionStepJSON]:
        return [ActionStepJSON(**step) for step in self.steps_json or []]

    @property
    def team(self) -> "TeamDataClass":
        """Placeholder for team relationship - would need to be injected"""
        return TeamDataClass(id=self.team_id, project_id=self.project_id)


@dataclass
class PropertyDefinitionDataClass:
    name: str
    type: PropertyDefinitionType
    property_type: Optional[PropertyType] = None
    team_id: Optional[int] = None
    project_id: Optional[int] = None
    group_type_index: Optional[int] = None

    @property
    def effective_project_id(self) -> Optional[int]:
        return self.project_id or self.team_id

    class Type:
        EVENT = PropertyDefinitionType.EVENT
        PERSON = PropertyDefinitionType.PERSON
        GROUP = PropertyDefinitionType.GROUP
        SESSION = PropertyDefinitionType.SESSION


@dataclass
class InsightVariableDataClass:
    id: Union[str, UUID]
    team_id: int
    name: str
    type: str
    code_name: Optional[str] = None
    default_value: Any = None

    @property
    def pk(self) -> Union[str, UUID]:
        return self.id


@dataclass
class ElementDataClass:
    """Element data with useful elements constant"""
    
    USEFUL_ELEMENTS = ["a", "button", "input", "select", "textarea", "label"]
    
    tag_name: Optional[str] = None
    text: Optional[str] = None
    href: Optional[str] = None
    attr_id: Optional[str] = None
    attr_class: Optional[list[str]] = None


# Re-export commonly used UUID types
class UUIDType:
    """Placeholder for UUIDT functionality"""
    
    @staticmethod
    def generate() -> str:
        from uuid import uuid4
        return str(uuid4())