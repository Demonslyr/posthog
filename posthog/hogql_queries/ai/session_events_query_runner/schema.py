"""
Schema definitions for SessionBatchEventsQuery extending EventsQuery.

This approach leverages PostHog's existing EventsQuery infrastructure while adding
multi-session capabilities for session summary workflows.
"""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from ee.session_recordings.session_summary.input_data import EXTRA_SUMMARY_EVENT_FIELDS
from posthog.schema import (
    CachedEventsQueryResponse,
    EventsQuery,
    EventsQueryResponse,
)
from posthog.session_recordings.queries.session_replay_events import DEFAULT_EVENT_FIELDS


class SessionBatchEventsQuery(EventsQuery):
    """
    Extended EventsQuery for fetching events from multiple sessions efficiently.

    This extends the standard EventsQuery with session-specific capabilities:
    - Batch querying multiple sessions in a single request
    - Session-grouped result organization

    Inherits all EventsQuery functionality:
    - Standard field selection via `select`
    - Property filtering via `properties` and `fixedProperties`
    - Time filtering via `after` and `before`
    - HogQL filtering via `where`
    - Ordering via `orderBy`
    - Pagination via `limit` and `offset`
    """

    model_config = ConfigDict(extra="forbid")

    # Override kind to distinguish from base EventsQuery
    kind: Literal["SessionBatchEventsQuery"] = "SessionBatchEventsQuery"

    # Session-specific fields
    session_ids: list[str] = Field(
        ..., description="List of session IDs to fetch events for. Will be translated to $session_id IN filter."
    )

    group_by_session: bool = Field(default=True, description="Whether to group results by session_id in the response")

    # Override response type to use our extended response
    response: Optional[SessionBatchEventsQueryResponse] = None


class SessionEventsItem(BaseModel):
    """
    Events from a single session within a batch query result.
    """

    model_config = ConfigDict(extra="forbid")

    session_id: str = Field(..., description="Session ID these events belong to")

    events: list[list[Any]] = Field(
        ...,
        description="List of events for this session, each event is a list of field values matching the query columns",
    )


class SessionBatchEventsQueryResponse(EventsQueryResponse):
    """Extended EventsQueryResponse for session batch queries."""

    model_config = ConfigDict(extra="forbid")

    # Session-grouped results (when group_by_session=True)
    session_events: Optional[list[SessionEventsItem]] = Field(
        default=None, description="Events grouped by session ID. Only populated when group_by_session=True."
    )

    sessions_with_no_events: list[str] = Field(
        default_factory=list, description="List of session IDs that had no matching events"
    )


class CachedSessionBatchEventsQueryResponse(CachedEventsQueryResponse):
    """Cached version of SessionBatchEventsQueryResponse."""

    model_config = ConfigDict(extra="forbid")

    session_events: Optional[list[SessionEventsItem]] = Field(
        default=None, description="Events grouped by session ID. Only populated when group_by_session=True."
    )
    sessions_with_no_events: list[str] = Field(
        default_factory=list, description="List of session IDs that had no matching events"
    )


# Type alias for convenience
SessionEventsResults = dict[str, list[list[Any]]]  # session_id -> events mapping


def create_session_batch_events_query(
    session_ids: list[str],
    select: Optional[list[str]] = None,
    events_to_ignore: Optional[list[str]] = None,
    after: Optional[str] = None,
    before: Optional[str] = None,
    max_total_events: Optional[int] = None,
    offset: Optional[int] = None,
    include_session_id: bool = True,
    **kwargs: Any,
) -> SessionBatchEventsQuery:
    """Create query for getting events for multiple sessions"""

    # Default field selection for session summaries
    if select is None:
        select = DEFAULT_EVENT_FIELDS + EXTRA_SUMMARY_EVENT_FIELDS

    # Ensure $session_id is included for grouping
    if include_session_id and "properties.$session_id" not in select:
        select.append("properties.$session_id")

    # Build WHERE clauses for session filtering and event exclusion
    where_clauses = [f"properties.$session_id IN {session_ids}"]  # Filter by session IDs by default

    # Exclude unwanted events (default to ignoring feature flag calls)
    if events_to_ignore is None:
        events_to_ignore = ["$feature_flag_called"]
    if events_to_ignore:
        event_list = "', '".join(events_to_ignore)
        where_clauses.append(f"event NOT IN ('{event_list}')")

    # Combine with any existing where clauses, if provided
    existing_where = kwargs.get("where", [])
    if isinstance(existing_where, list):
        where_clauses.extend(existing_where)

    # Set defaults for session batch queries
    query_params = {
        "select": select,
        "session_ids": session_ids,
        "where": where_clauses,
        "orderBy": ["properties.$session_id", "timestamp ASC"],  # Group by session, then chronological
        "limit": max_total_events,
        "offset": offset,
        "group_by_session": True,
        "after": after,
        "before": before,
        **kwargs,  # Allow overriding defaults
    }
    return SessionBatchEventsQuery(**query_params)
