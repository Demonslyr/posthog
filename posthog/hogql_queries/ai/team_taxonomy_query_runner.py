from posthog.hogql import ast
from posthog.hogql.parser import parse_select
from posthog.hogql.printer import to_printed_hogql
from posthog.hogql.query import execute_hogql_query
from posthog.hogql_queries.ai.utils import TaxonomyCacheMixin
from posthog.hogql_queries.query_runner import QueryRunner
from posthog.schema import (
    CachedTeamTaxonomyQueryResponse,
    TeamTaxonomyItem,
    TeamTaxonomyQuery,
    TeamTaxonomyQueryResponse,
)
from difflib import SequenceMatcher
from typing import Optional, cast
from ee.models.event_definition import EnterpriseEventDefinition
from nltk import ngrams
from collections import Counter
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from functools import lru_cache
from scipy.sparse import spmatrix

try:
    from posthog.taxonomy.taxonomy import CORE_FILTER_DEFINITIONS_BY_GROUP
except ImportError:
    CORE_FILTER_DEFINITIONS_BY_GROUP = {}


class CachedTfidfVectorizer:
    def __init__(self):
        self._vectorizer = TfidfVectorizer()
        self._last_texts = None
        self._last_matrix = None

    def fit_transform(self, texts: list[str]) -> spmatrix:
        if self._last_texts == texts and self._last_matrix is not None:
            return self._last_matrix

        self._last_texts = texts
        self._last_matrix = self._vectorizer.fit_transform(texts)
        return self._last_matrix


# Global instance
_vectorizer = CachedTfidfVectorizer()


def get_event_descriptions_batch(event_names: list[str]) -> dict[str, Optional[str]]:
    """Get event descriptions in batch from either core definitions or enterprise definitions"""
    result: dict[str, Optional[str]] = {}

    # Get from core definitions
    for event_name in event_names:
        if event_core_definition := CORE_FILTER_DEFINITIONS_BY_GROUP.get("events", {}).get(event_name):
            label = event_core_definition.get("label_llm") or event_core_definition.get("label")
            if label:
                description = f"{label}. {event_core_definition.get('description', '')}"
            else:
                description = event_core_definition.get("description", "")
            result[event_name] = description

    # Get remaining from EnterpriseEventDefinition
    remaining_events = {e for e in event_names if e not in result}
    if remaining_events:
        try:
            enterprise_defs = EnterpriseEventDefinition.objects.filter(name__in=remaining_events)
            for def_ in enterprise_defs:
                if def_.description:
                    result[def_.name] = def_.description
        except Exception:
            pass

    # Set None for events without descriptions
    for event_name in event_names:
        if event_name not in result:
            result[event_name] = None

    return result


@lru_cache(maxsize=1000)
def calculate_sequence_similarity(text1: str, text2: str) -> float:
    """Calculate similarity using SequenceMatcher with caching"""
    return SequenceMatcher(None, text1.lower(), text2.lower()).ratio()


@lru_cache(maxsize=1000)
def calculate_ngram_similarity(text1: str, text2: str, n: int = 2) -> float:
    """Calculate similarity using n-grams and Jaccard similarity with caching"""
    text1 = text1.lower()
    text2 = text2.lower()

    # Generate n-grams
    ngrams1 = Counter(ngrams(text1.split(), n))
    ngrams2 = Counter(ngrams(text2.split(), n))

    # Calculate Jaccard similarity
    intersection = sum((ngrams1 & ngrams2).values())
    union = sum((ngrams1 | ngrams2).values())

    return intersection / union if union > 0 else 0.0


def calculate_tfidf_similarity_batch(query: str, texts: list[str]) -> list[float]:
    """Calculate TF-IDF similarity for multiple texts at once"""
    if not texts:
        return []

    all_texts = [query.lower()] + [t.lower() for t in texts]
    tfidf_matrix = _vectorizer.fit_transform(all_texts)

    # Convert sparse matrix to dense array for similarity calculation
    query_matrix = tfidf_matrix[0:1]
    text_matrix = tfidf_matrix[1:]
    similarities = cosine_similarity(query_matrix, text_matrix)

    # Convert to list of floats
    return cast(list[float], similarities[0].tolist())


def calculate_similarity_batch(
    query: str, event_names: list[str], descriptions: dict[str, Optional[str]]
) -> dict[str, float]:
    """Calculate similarity scores for multiple events at once"""
    sequence_similarities = [calculate_sequence_similarity(query, name) for name in event_names]
    ngram_similarities = [calculate_ngram_similarity(query, name) for name in event_names]
    tfidf_similarities = calculate_tfidf_similarity_batch(query, event_names)

    # Adjust similarities based on descriptions
    for i, event_name in enumerate(event_names):
        if desc := descriptions[event_name]:
            desc_sequence = calculate_sequence_similarity(query, desc)
            desc_ngram = calculate_ngram_similarity(query, desc)
            sequence_similarities[i] = max(sequence_similarities[i], desc_sequence)
            ngram_similarities[i] = max(ngram_similarities[i], desc_ngram)

    # Weight and combine similarities
    weights = {"sequence": 0.3, "ngram": 0.4, "tfidf": 0.3}

    return {
        event_name: weights["sequence"] * seq + weights["ngram"] * ngram + weights["tfidf"] * tfidf
        for event_name, seq, ngram, tfidf in zip(
            event_names, sequence_similarities, ngram_similarities, tfidf_similarities
        )
    }


class TeamTaxonomyQueryRunner(TaxonomyCacheMixin, QueryRunner):
    """
    Calculates the top events for a team sorted by count. The EventDefinition model doesn't store the count of events,
    so this query mitigates that.
    """

    query: TeamTaxonomyQuery
    response: TeamTaxonomyQueryResponse
    cached_response: CachedTeamTaxonomyQueryResponse

    def calculate(self):
        query = self.to_query()
        hogql = to_printed_hogql(query, self.team)

        response = execute_hogql_query(
            query_type="TeamTaxonomyQuery",
            query=query,
            team=self.team,
            timings=self.timings,
            modifiers=self.modifiers,
            limit_context=self.limit_context,
        )

        # Filter out system/ignored events
        filtered_results = [
            (event, count)
            for event, count in response.results
            if not (event_core_definition := CORE_FILTER_DEFINITIONS_BY_GROUP.get("events", {}).get(event))
            or not (event_core_definition.get("system") or event_core_definition.get("ignored_in_assistant"))
        ]

        # Get all event names for batch processing
        event_names: list[str] = list({event for event, _ in filtered_results})

        # Get descriptions in batch
        descriptions = get_event_descriptions_batch(event_names)

        # Calculate similarities in batch if query plan exists
        similarities = None
        if self.query.plan:
            similarities = calculate_similarity_batch(self.query.plan, event_names, descriptions)

        # Create results
        results: list[TeamTaxonomyItem] = []
        for event, count in filtered_results:
            results.append(
                TeamTaxonomyItem(
                    event=event,
                    count=count,
                    description=descriptions[event],
                    similarity=similarities[event] if similarities is not None else None,
                )
            )

        # Sort by similarity and count
        results.sort(key=lambda x: (-x.similarity if x.similarity is not None else 0, -x.count))

        return TeamTaxonomyQueryResponse(
            results=results, timings=response.timings, hogql=hogql, modifiers=self.modifiers
        )

    def to_query(self) -> ast.SelectQuery | ast.SelectSetQuery:
        query = parse_select(
            """
                SELECT
                    event,
                    count() as count
                FROM events
                WHERE
                    timestamp >= now () - INTERVAL 30 DAY
                GROUP BY
                    event
                ORDER BY
                    count DESC
                LIMIT 500
            """
        )

        return query
