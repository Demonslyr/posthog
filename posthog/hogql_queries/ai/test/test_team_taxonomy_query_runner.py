from datetime import timedelta

from django.test import override_settings
from django.utils import timezone
from freezegun import freeze_time

from posthog.hogql_queries.ai.team_taxonomy_query_runner import TeamTaxonomyQueryRunner
from posthog.schema import CachedTeamTaxonomyQueryResponse, TeamTaxonomyQuery
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
    _create_event,
    _create_person,
    flush_persons_and_events,
    snapshot_clickhouse_queries,
)
from ee.models.event_definition import EnterpriseEventDefinition


@override_settings(IN_UNIT_TESTING=True)
class TestTeamTaxonomyQueryRunner(ClickhouseTestMixin, APIBaseTest):
    @snapshot_clickhouse_queries
    def test_taxonomy_query_runner(self):
        _create_person(
            distinct_ids=["person1"],
            properties={"email": "person1@example.com"},
            team=self.team,
        )
        _create_event(
            event="event1",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        _create_event(
            event="event2",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        _create_event(
            event="event1",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )

        results = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery()).calculate()
        self.assertEqual(len(results.results), 2)
        self.assertEqual(results.results[0].event, "event1")
        self.assertEqual(results.results[0].count, 2)
        self.assertEqual(results.results[1].event, "event2")
        self.assertEqual(results.results[1].count, 1)

    @snapshot_clickhouse_queries
    def test_similarity_search(self):
        _create_person(
            distinct_ids=["person1"],
            properties={"email": "person1@example.com"},
            team=self.team,
        )
        _create_event(
            event="user_signup",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        _create_event(
            event="user_login",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        _create_event(
            event="page_view",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )

        # Test with a query plan that should match user_signup
        results = TeamTaxonomyQueryRunner(
            team=self.team, query=TeamTaxonomyQuery(plan="show users who signed up over the last month")
        ).calculate()

        self.assertEqual(len(results.results), 3)  # All events should be returned
        self.assertEqual(results.results[0].event, "user_signup")  # Should be first due to similarity
        self.assertIsNotNone(results.results[0].similarity)
        similarity = results.results[0].similarity
        assert similarity is not None  # Type assertion
        self.assertGreater(similarity, 0.1)

        # Test with a query plan that should match page_view
        results = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery(plan="show page views")).calculate()

        self.assertEqual(len(results.results), 3)  # All events should be returned
        self.assertEqual(results.results[0].event, "page_view")  # Should be first due to similarity
        self.assertIsNotNone(results.results[0].similarity)
        similarity = results.results[0].similarity
        assert similarity is not None  # Type assertion
        self.assertGreater(similarity, 0.1)

    @snapshot_clickhouse_queries
    def test_similarity_batch_calculation(self):
        _create_person(distinct_ids=["person1"], team=self.team)
        events = ["user_signup", "user_login", "page_view", "button_click", "form_submit"]
        for event in events:
            _create_event(event=event, distinct_id="person1", team=self.team)

        results = TeamTaxonomyQueryRunner(
            team=self.team, query=TeamTaxonomyQuery(plan="user authentication events")
        ).calculate()

        # Should prioritize auth-related events
        first_similarity = results.results[0].similarity
        last_similarity = results.results[-1].similarity
        assert first_similarity is not None and last_similarity is not None
        self.assertGreater(first_similarity, last_similarity)
        self.assertIn(results.results[0].event, ["user_signup", "user_login"])

    @snapshot_clickhouse_queries
    def test_description_similarity(self):
        _create_person(distinct_ids=["person1"], team=self.team)
        _create_event(event="custom_event", distinct_id="person1", team=self.team)

        # Mock enterprise event definition with description
        EnterpriseEventDefinition.objects.create(
            team=self.team, name="custom_event", description="User completed onboarding flow"
        )

        results = TeamTaxonomyQueryRunner(
            team=self.team, query=TeamTaxonomyQuery(plan="onboarding completion")
        ).calculate()

        self.assertEqual(results.results[0].event, "custom_event")
        similarity = results.results[0].similarity
        assert similarity is not None
        self.assertGreater(similarity, 0.1)

    @snapshot_clickhouse_queries
    def test_empty_similarity_inputs(self):
        _create_person(distinct_ids=["person1"], team=self.team)
        _create_event(event="event1", distinct_id="person1", team=self.team)

        # Empty query plan
        results = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery(plan="")).calculate()

        self.assertEqual(len(results.results), 1)
        self.assertIsNone(results.results[0].similarity)

        # Very short query
        results = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery(plan="a")).calculate()

        self.assertEqual(len(results.results), 1)
        self.assertIsNotNone(results.results[0].similarity)

    @snapshot_clickhouse_queries
    def test_similarity_weights(self):
        _create_person(distinct_ids=["person1"], team=self.team)
        events = [
            ("user_signup", "User signed up for the first time"),
            ("signup", "User created an account"),
            ("new_user", "New user registration"),
        ]
        for event, desc in events:
            _create_event(event=event, distinct_id="person1", team=self.team)
            EnterpriseEventDefinition.objects.create(team=self.team, name=event, description=desc)

        results = TeamTaxonomyQueryRunner(
            team=self.team, query=TeamTaxonomyQuery(plan="new user registration")
        ).calculate()

        # Should consider both name and description similarity
        similarity = results.results[0].similarity
        assert similarity is not None
        self.assertGreater(similarity, 0.5)

    @snapshot_clickhouse_queries
    def test_empty_text_list(self):
        _create_person(distinct_ids=["person1"], team=self.team)
        _create_event(event="event1", distinct_id="person1", team=self.team)

        # Test with empty event list
        results = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery(plan="test")).calculate()

        self.assertEqual(len(results.results), 1)
        self.assertIsNotNone(results.results[0].similarity)

    def test_caching(self):
        now = timezone.now()

        with freeze_time(now):
            _create_person(
                distinct_ids=["person1"],
                properties={"email": "person1@example.com"},
                team=self.team,
            )
            _create_event(
                event="event1",
                distinct_id="person1",
                team=self.team,
            )

            runner = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery())
            response = runner.run()

            assert isinstance(response, CachedTeamTaxonomyQueryResponse)
            self.assertEqual(len(response.results), 1)

            key = response.cache_key
            _create_event(
                event="event2",
                distinct_id="person1",
                team=self.team,
            )
            flush_persons_and_events()

            runner = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery())
            response = runner.run()

            assert isinstance(response, CachedTeamTaxonomyQueryResponse)
            self.assertEqual(response.cache_key, key)
            self.assertEqual(len(response.results), 1)

        with freeze_time(now + timedelta(minutes=59)):
            runner = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery())
            response = runner.run()

            assert isinstance(response, CachedTeamTaxonomyQueryResponse)
            self.assertEqual(len(response.results), 1)

        with freeze_time(now + timedelta(minutes=61)):
            runner = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery())
            response = runner.run()

            assert isinstance(response, CachedTeamTaxonomyQueryResponse)
            self.assertEqual(len(response.results), 2)

    def test_limit(self):
        now = timezone.now()

        _create_person(
            distinct_ids=["person1"],
            properties={"email": "person1@example.com"},
            team=self.team,
        )

        for i in range(501):
            with freeze_time(now + timedelta(minutes=i)):
                _create_event(
                    event=f"event{i}",
                    distinct_id="person1",
                    team=self.team,
                )

        flush_persons_and_events()

        runner = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery())
        response = runner.run()

        assert isinstance(response, CachedTeamTaxonomyQueryResponse)
        self.assertEqual(len(response.results), 500)

    def test_events_not_useful_for_llm_ignored(self):
        _create_person(
            distinct_ids=["person1"],
            properties={"email": "person1@example.com"},
            team=self.team,
        )
        for _i in range(2):
            _create_event(
                event="$pageview",
                distinct_id="person1",
                properties={"$browser": "Chrome", "$country": "US"},
                team=self.team,
            )
        _create_event(
            event="did custom thing",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        # Events below should have `ignored_in_assistant`
        _create_event(
            event="$pageleave",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        _create_event(
            event="$autocapture",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )
        _create_event(
            event="$feature_flag_called",
            distinct_id="person1",
            properties={"$browser": "Chrome", "$country": "US"},
            team=self.team,
        )

        flush_persons_and_events()

        runner = TeamTaxonomyQueryRunner(team=self.team, query=TeamTaxonomyQuery())
        response = runner.run()

        assert isinstance(response, CachedTeamTaxonomyQueryResponse)
        self.assertEqual([result.event for result in response.results], ["$pageview", "did custom thing"])
