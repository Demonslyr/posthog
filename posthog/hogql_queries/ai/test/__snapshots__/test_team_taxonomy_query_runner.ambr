# serializer version: 1
# name: TestTeamTaxonomyQueryRunner.test_description_similarity
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_empty_similarity_inputs
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_empty_similarity_inputs.1
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_empty_text_list
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_similarity_batch_calculation
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_similarity_search
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_similarity_search.1
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_similarity_weights
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestTeamTaxonomyQueryRunner.test_taxonomy_query_runner
  '''
  SELECT events.event AS event,
         count() AS count
  FROM events
  WHERE and(equals(events.team_id, 99999), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'), minus(now64(6, 'UTC'), toIntervalDay(30))))
  GROUP BY events.event
  ORDER BY count DESC
  LIMIT 500 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
